src/feature/hs/hs_service.c

```diff
@@ -2364,9 +2373,7 @@ cleanup_intro_points(hs_service_t *service, time_t now)
        * the node_t anymore (removed from our latest consensus) or if we've
        * reached the maximum number of retry with a non existing circuit
        * and don't have any established circuits. */
-      if (has_expired || node == NULL ||
-          (ip->circuit_retries > MAX_INTRO_POINT_CIRCUIT_RETRIES &&
-           (!ip->circuit_established || hs_circ_service_get_intro_circ(ip)))) {
+      if (has_expired || node == NULL || can_retry_intro_point(ip)) {
```

Comment:
I'm a bit confused on why we trash the intro point if we can retry it. Intuitively we should do the opposite right?